docs/developers/debugging/compile_time_regressions.md

```diff
@@ -0,0 +1,225 @@
+# Debugging Compile Time Regressions
+
+So the IREE compiler used to compile a program quickly, but it is now slower.
+What do you do?
+
+## Initial information gathering
+
+Try to answer as many of these questions as you can:
+
+* **When did compilation get slower?**
+
+  A specific git commit is ideal, but "sometime in the last week" is a good
+  starting point. You'll ultimately want to find a culprit release or git
+  commit that changed the compiler code.
+
+* **How much slower did compilation get?**
+
+  Be specific - did it jump from 1 minute to 2 minutes, or 1 minute to 1 hour?
+  Identifying the scale of the regression can help set the priority to
+  investigate it.
+
+* **What is the full compile command?**
+
+  Try to extract the input program and full list of flags passed to the
+  compiler binary so that others can reproduce what you're seeing. Try to
+  distill this as much as possible to using just native tools (no Python or
+  other framework layers).
+
+* **What environment is the compiler running in?**
+
+  Are you using a `Debug` build, or a release build? What operating system and
+  size machine is running the compiler (e.g. Linux developer machine, or a
+  smaller system)?
+
+## Culprit finding and bisecting
+
+If you only have a rough idea of when something changed and want to narrow that
+down to a specific code change, bisecting can help.
+
+### Running `git bisect`
+
+Building the compiler from source and using
+[`git bisect`](https://git-scm.com/docs/git-bisect) will let you pinpoint
+specific commits in IREE, though it typically won't let you step through changes
+in submodules (e.g. MLIR updates in `third_party/llvm-project/`).
+
+__Tip__: [Configure ccache](../developing_iree/ccache.md) if you'll be rebuilding the compiler while bisecting
+
+A manual workflow with `git bisect` looks like this:
+
+```bash
+git bisect start --first-parent
+git bisect good [<rev>]
+git bisect bad [<rev>]
+
+# Read the prompts from the command as it runs
+# At each step, test the compiler:
+#   git submodule update
+#   cmake --build build/ --target iree-compile
+#   ./build/tools/iree-compile <args>
+#       attach Tracy, observe timing, print IR, etc. to determine if fast or slow
+#       if fast, `git bisect good`
+#       if slow, `git bisect bad`
+#   repeat
+```
+
+An automated workflow can use `git bisect run` and a script:
+
+```shell
+# run_bisect.sh
+git submodule update
+cmake --build build/ --target iree-compile
```

Comment:
FWIW, this is the script I used to deal with submodules

```shell
#!/bin/bash

set -uo pipefail

function update_submodules() {
  # The only way to actually clear out old submodules reliably...
  rm -rf third_party/
  git restore .
  git submodule sync
  git submodule update --init
  git clean -f .
  git add -A
  git diff HEAD --exit-code
}

(set -e; update_submodules) || exit -1 
"$@"
```

I don't remember why I chose `-1 = 255 (mod 256)` exactly other than being an exit code that would cause an abort

Reply:
Thanks. I considered at least running `git submodule sync`, but decided to keep the sample script focused on the bisect use case, avoiding some of the edge cases that can happen with the git repo. Some of those operations add a bit too much overhead (both in time spent and in logs generated) while on the happy path for my taste.